system/jlib/jexcept.cpp

```diff
@@ -980,19 +980,45 @@ static void throwSigSegV()
 }
 #endif
 
+void termhandler(int signum)
+{
+    // allow time for flush in excsighandler to finish
+    Sleep(12500);
+    _exit(0);
+}
+
 void excsighandler(int signum, siginfo_t *info, void *extra) 
 {
     static byte nested=0;
     if (nested++)
         return;
 
     excsignal = 0;
-#if defined(NO_LINUX_SEH) && !defined(SA_RESETHAND)
-    signal(SIGSEGV, SIG_DFL);
-    signal(SIGBUS, SIG_DFL);
-    signal(SIGILL, SIG_DFL);
-    signal(SIGFPE, SIG_DFL);
-    signal(SIGABRT, SIG_DFL);
+#if defined(NO_LINUX_SEH)
+    sigset_t blockset;
+    sigemptyset(&blockset);
+
+    struct sigaction act;
+    act.sa_mask = blockset;
+    act.sa_flags = 0;
+    act.sa_handler = SIG_DFL;
+    sigaction(SIGSEGV, &act, NULL);
+    sigaction(SIGILL, &act, NULL);
+    sigaction(SIGBUS, &act, NULL);
+    sigaction(SIGFPE, &act, NULL);
+    sigaction(SIGABRT, &act, NULL);
+
+    // don't get pre-empted by inits' as we are ending soon
+    sigaddset(&blockset, SIGINT);
+    sigaddset(&blockset, SIGQUIT);
+    sigaddset(&blockset, SIGTERM);
+    struct sigaction act2;
+    act2.sa_mask = blockset;
+    act2.sa_flags = 0;
+    act2.sa_handler = &termhandler;
+    sigaction(SIGINT, &act2, NULL);
+    sigaction(SIGQUIT, &act2, NULL);
+    sigaction(SIGTERM, &act2, NULL);
```

Comment:
Is it safe to install a signal handler whilst already inside a signal handler?

It's also not clear to me why this is really necessary or if the constant 12.5 second delay is sensible.
Again, may be just a case of requiring fuller commented explanation, so reader can understand logic/reasoning here.